[v3 image assets] (#3739)
withastro/docs: src/content/docs/en/guides/images.mdx

i18nReady: true
---
import Since from '~/components/Since.astro';
import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro';

> Astro provides several ways for you to use images on your site, whether they are stored locally inside your project, linked to remotely, or stored in a CMS or CDN!

Reviewer: I guess this was already here, but "linked to remotely" sounds off? I would even say that remote images are the same thing as using with a  CMS / CDN? Is there any technical difference that I am missing?

```suggestion
Astro provides several ways for you to use images on your site, whether they are stored locally inside your project, or linked remotely through a CMS or CDN!
```

Author: Yeah, I felt that too this time around. BUT, there is a random, one-off inline `src="url"` vs using a CMS/CDN. They aren't the same thing so I don't want to make it sound like you can't just use a random URL. I'll think about it while I apply the Wrath of the Commas.

> Astro provides several ways for you to use images on your site, whether they are stored locally inside your project, linked to remotely, or stored in a CMS or CDN!

Author: ```suggestion
Astro provides several ways for you to use images on your site, whether they are stored locally inside your project, linked to from an external URL, or stored in a CMS or CDN!
```

Author: @Yan-Thomas is this any better?  😅

Reviewer: LGTM!

## Where to store images

### `src/` vs `public/`

We recommend that local images are kept in `src/` when possible so that Astro can transform, optimize and bundle them. Files in the `/public` directory are always served or copied into the build folder as-is, with no processing. 

> Your local images stored in `src/` can be used by all files in your project: `.astro`, `.md`, `.mdx`, `.mdoc` and other UI frameworks. Images can be stored in any folder, including alongside your content. 

Reviewer: Comma I - The Phantom Menace
```suggestion
Your local images stored in `src/` can be used by all files in your project: `.astro`, `.md`, `.mdx`, `.mdoc`, and other UI frameworks. Images can be stored in any folder, including alongside your content.
```

## Images in `.astro` files

In `.astro` files, local images must be imported into the file in order to be used. Remote and `public/` images do not require importing.

Import and use Astro's built-in `<Image />` component for optimized images using `astro:assets`. Alternatively, Astro syntax supports writing an HTML `<img>` tag directly, which skips image processing. 

```astro title="src/pages/blog/MyImages.astro"
---
import { Image } from 'astro:assets';
import localBirdImage from ''../../images/subfolder/localBirdImage.png';
---
<Image src={localBirdImage} alt="A bird sitting on a nest of eggs."/>
<Image src="/bird-public-folder.jpg" alt="A bird." width="50" height="50"/>
<Image src="https://example.com/remote-bird.jpg" alt="A bird." width="50" height="50"/>

<img src={localBirdImage.src} alt="A bird sitting on a nest of eggs.">
<img src="/images/bird-in-public-folder.jpg" alt="A bird.">
<img src="https://example.com/remote-bird.jpg" alt="A bird.">
```

### `<Image />` (`astro:assets`)

Use the built-in `<Image />` Astro component to display optimized versions of your local images and [configured remote images](#authorizing-remote-images).

> Images in the `public/` folder, as well as remote images not specifically configured in your project can also be used with the Image component, but will not be processed.

Reviewer: Comma II - Attack of the Clones
```suggestion
Images in the `public/` folder, as well as remote images not specifically configured in your project, can also be used with the Image component, but will not be processed.
```

> `<Image />` can transform a local or authorized remote image's dimensions, file type, and quality for control over your displayed image. The resulting `<img>` tag includes `alt`, `loading` and `decoding` attributes and infers image dimensions to avoid **Cumulative Layout Shift (CLS)**.

Reviewer: Comma III - Revenge of the Sith
````suggestion
`<Image />` can transform a local or authorized remote image's dimensions, file type, and quality for control over your displayed image. The resulting `<img>` tag includes `alt`, `loading`, and `decoding` attributes and infers image dimensions to avoid **Cumulative Layout Shift (CLS)**.
````

  ---
  import { Image } from 'astro:assets';
  ---
  <Image 
    src="https://example.com/remote-image.jpg"
    alt="descriptive text"
    width="200"
    height="150" />
  ```

##### alt (required)

Use the required `alt` attribute to provide a string of [descriptive alt text](https://www.w3.org/WAI/tutorials/images/) for images.

If an image is merely decorative (i.e. doesn't contribute to the understanding of the page), set `alt=""` so that screen readers and other assistive technologies know to ignore the image.

##### width and height (required for `public/` and remote images)

These properties define the dimensions to use for the image.

When using local images in their original aspect ratio, the `width` and `height` can be automatically inferred from the source file and are optional.

However, both of these properties are required for remote images and images stored in your `public/` folder as Astro is unable to analyze these files.

##### format 

You can optionally state the [image file type](https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Image_types#common_image_file_types) output to be used.

> By default, the `<Image />` component will produce a `webp` file.

Reviewer: When we mention other formats we always add the dot in the start, shouldn't we follow this for image formats as well?
```suggestion
By default, the `<Image />` component will produce a `.webp` file.
```

# My Markdown Page

<!-- Local images now possible! -->
![A starry night sky.](../../images/stars.png)

<!-- Keep your images next to your content! -->
![A starry night sky.](./stars.png)
```

If you require more control over your image attributes, we recommend using the `.mdx` file format, which allows you to include Astro's `<Image />` component or a JSX `<img />` tag in addition to the Markdown syntax. Use the [MDX integration](/en/guides/integrations-guide/mdx/) to add support for MDX to Astro.

### Remove `@astrojs/image`

If you were using the image integration in Astro v2.x, complete the following steps:

1. Remove the `@astrojs/image` integration.

    You must [remove the integration](/en/guides/integrations-guide/#removing-an-integration) by uninstalling and then removing it from your `astro.config.mjs` file.

2. Migrate any existing `<Image />` components.

    Change all `import` statements from `@astrojs/image/components` to `astro:assets` in order to use the new built-in `<Image />` component.

    Remove any component attributes that are not [currently supported image asset properties](/en/guides/images/#properties).

>     For example `aspectRatio` is no longer supported, as it is now automatically inferred from the `width` and `height` attributes.

Reviewer: Comma IV - A New Hope
```suggestion
    For example, `aspectRatio` is no longer supported, as it is now automatically inferred from the `width` and `height` attributes.
```

Your local images stored in `src/` can be used by all files in your project: `.astro`, `.md`, `.mdx`, `.mdoc` and other UI frameworks. Images can be stored in any folder, including alongside your content. 

Store your images in the `public/` folder only if you want to avoid any processing.

### Remote images

You can also choose to store your images remotely, in a content management system (CMS) or digital asset management (DAM) platform. 

For extra protection when dealing with external sources, remote images will only be processed from [authorized image sources](#authorizing-remote-images) specified in your configuration. However, any remote images can be displayed.

Astro can fetch your data remotely using APIs or display images from their full URL path. See our [CMS guides](/en/guides/cms/) for examples of integrating common services.

## Images in `.astro` files

In `.astro` files, local images must be imported into the file in order to be used. Remote and `public/` images do not require importing.

> Import and use Astro's built-in `<Image />` component for optimized images. Alternatively, Astro syntax supports writing an HTML `<img>` tag directly.

Reviewer 2: > Alternatively, Astro syntax supports writing an HTML `<img>` tag directly.

Maybe it should be specified that this come with no processing?

> Import and use Astro's built-in `<Image />` component for optimized images. Alternatively, Astro syntax supports writing an HTML `<img>` tag directly.

Author: ```suggestion
Import and use Astro's built-in `<Image />` component for optimized images using `astro:assets`. Alternatively, Astro syntax supports writing an HTML `<img>` tag directly, which skips image processing. 
```

```astro title="src/pages/blog/MyImages.astro"
---
import { Image } from 'astro:assets';
import localBirdImage from ''../../images/subfolder/localBirdImage.png';
---
<Image src={localBirdImage} alt="A bird sitting on a nest of eggs."/>
<Image src="/bird-public-folder.jpg" alt="A bird." width="50" height="50"/>
<Image src="https://example.com/remote-bird.jpg" alt="A bird." width="50" height="50"/>

<img src={localBirdImage.src} alt="A bird sitting on a nest of eggs.">
<img src="/images/bird-in-public-folder.jpg" alt="A bird.">
<img src="https://example.com/remote-bird.jpg" alt="A bird.">
```

### `<Image />` (`astro:assets`)

Use the built-in `<Image />` Astro component to display optimized versions of your local images and [configured remote images](#authorizing-remote-images).

Images in the `public/` folder, as well as remote images not specifically configured in your project can also be used with the Image component, but will not be processed.

> `<Image />` can transform a local or authorized remote image's dimensions, file type, and quality to produce an optimized `<img>` tag to avoid **Cumulative Layout Shift (CLS)**.

Reviewer 2: This sentence felt a bit clunky to me, an "optimized `img` tag" doesn't mean much and felt weird considering previous paragraphs refers to "optimized images". 

Additionally, while the CLS is the main part, other benefits of the `img` tag generated is enforcement of `alt` and attributes such as `loading` and `decoding`.

> `<Image />` can transform a local or authorized remote image's dimensions, file type, and quality to produce an optimized `<img>` tag to avoid **Cumulative Layout Shift (CLS)**.

Author: ```suggestion
`<Image />` can transform a local or authorized remote image's dimensions, file type, and quality for control over your displayed image. The resulting `<img>` tag includes `alt`, `loading` and `decoding` attributes and infers image dimensions to avoid **Cumulative Layout Shift (CLS)**.
```

:::tip[What is Cumulative Layout Shift?]

> [Cumulative Layout Shift (CLS)](https://web.dev/cls/) is a Core Web Vital metric for measuring how much content shifted on your page during loading. The `<Image />` component optimizes for CLS by automatically selecting the correct `width` and `height` for your local images.

Reviewer 2: > selecting the correct `width` and `height` for your local images.

`select` might be the wrong verb here? It does not really select anything. Maybe just `outputting`?

> [Cumulative Layout Shift (CLS)](https://web.dev/cls/) is a Core Web Vital metric for measuring how much content shifted on your page during loading. The `<Image />` component optimizes for CLS by automatically selecting the correct `width` and `height` for your local images.

Author: ```suggestion
[Cumulative Layout Shift (CLS)](https://web.dev/cls/) is a Core Web Vital metric for measuring how much content shifted on your page during loading. The `<Image />` component optimizes for CLS by automatically setting the correct `width` and `height` for your local images.
```

:::

```astro title="src/components/MyComponent.astro"
---
// import the Image component and the image
import { Image } from 'astro:assets';
import myImage from "../assets/my_image.png"; // Image is 1600x900
---

<!-- `alt` is mandatory on the Image component -->
<Image src={myImage} alt="A description of my image." />
```

```html
<!-- Output -->
<!-- Image is optimized, proper attributes are enforced -->
<img
  src="/_astro/my_image.hash.webp"
  width="1600"
  height="900"
  decoding="async"
  loading="lazy"
  alt="A description of my image."
/>
```

Currently, the built-in assets feature does not include a `<Picture />` component.

> Instead, you can use the HTML image attributes `srcset` and `sizes` or the `<picture>` tag [for art direction or to create responsive images](https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images#art_direction).

Reviewer 2: You can do that, but since this doesn't teach you about `getImage` yet, from the user's perspective they don't know how to actually use those attributes and tag to use optimized images and multiple sizes at the same time. Using those attributes on `<Image />` won't do anything.

If an image is merely decorative (i.e. doesn't contribute to the understanding of the page), set `alt=""` so that screen readers and other assistive technologies know to ignore the image.

##### width and height (required for `public/` and remote images)

These properties define the dimensions to use for the image.

When using local images in their original aspect ratio, the `width` and `height` can be automatically inferred from the source file and are optional.

However, both of these properties are required for remote images and images stored in your `public/` folder as Astro is unable to analyze these files.

##### format 

You can optionally state the [image file type](https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Image_types#common_image_file_types) output to be used.

By default, the `<Image />` component will produce a `webp` file.

##### quality

`quality` is an optional property that can either be:
- a preset (`low`, `mid`, `high`, `max`) that is automatically normalized between formats.
- a number from `0` to `100` (interpreted differently between formats).

> We don't know what they mean, either.

Reviewer 2: ha

> We don't know what they mean, either.

Author: ```suggestion
```

For example, you can provide a `class` to the final `<img>` element.

```astro title="src/pages/index.astro"
---
import { Image } from 'astro:assets';
import myImage from "../assets/my_image.png";
---

<!-- `alt` is mandatory on the Image component -->
<Image src={myImage} alt="" class="my-class" />
```

```astro
<!-- Output -->
<img
  src="/_astro/my_image.hash.webp"
  width="1600"
  height="900"
  decoding="async"
  loading="lazy"
  class="my-class"
  alt=""
/>
```

#### Setting Default Values

Currently, there is no way to specify default values for all `<Image />` components. Required attributes should be set on each individual component.

As an alternative, you can wrap these components in another Astro component for reuse. For example, you could create a component for your blog post images:

```astro title="src/components/BlogPostImage.astro"
---
import { Image } from 'astro:assets';

const {src, ...attrs} = Astro.props;
---
<Image src={src} {...attrs} />

<style>
  img :global(img), svg {
    margin-block: 2.5rem;
    border-radius: 0.75rem;
  }
</style>
```

### `<img>`

> The [Astro template syntax](/en/core-concepts/astro-syntax/) also supports writing an `<img>` tag directly, with full control over its final output. Local images will be optimized.

Reviewer 2: > Local images will be optimized.

They won't! Only `getImage` and `Image` will process the image. Of course, you can pass `getImage`'s result to an `img` tag for it to show a processed image, but by itself it does nothing.

It accepts all HTML `<img>` tag properties, and the only required property is `src`.

#### Local images in `src/`

> Local images must be **imported from the relative path** from the existing `.astro` file, or configure and use an [import alias](/en/guides/aliases/). Then, you can access the image's `src` and other properties to use in the `<img>` tag.

Author: ```suggestion
Local images must be **imported from the relative path** from the existing `.astro` file, or configure and use an [import alias](/en/guides/aliases/). Then, you can access the image's `src` and other properties to use in the `<img>` tag.

For example, use the image's own `height` and `width` properties by default to avoid CLS and improve Core Web Vitals.
```

```astro "myDog.src"
---
// import local images
import myDog from `../../images/pets/local-dog.jpg`
---
// access the image properties

> <img src={myDog.src} alt="A barking dog." />

Reviewer 2: ```suggestion
<img src={myDog.src} width={myDog.width} height={myDog.height} alt="A barking dog." />
```

Shows the other attributes and show how you can use them to avoid CLS, it's neat imo

```

Imported image assets match the following signature:

```ts
interface ImageMetadata {
  src: string;
  width: number;
  height: number;
  format: string;
}
```

#### Images in `public/`
For images located within `public/` use the image's **file path relative to the public folder** as the `src` value:

```astro '"/images/public-cat.jpg"'
<img src="/images/public-cat.jpg" alt="A sleeping cat.">
```

#### Remote images

For remote images, use the image's **full URL** as the `src` value:

```astro '"https://example.com/remote-cat.jpg"'
<img src="https://example.com/remote-cat.jpg" alt="A sleeping cat.">
```

### Choosing `<Image />` vs `<img>`

The `<Image />` component optimizes your image and infers width and height (of local images) based on the original aspect ratio to avoid CLS. But, it only works with certain formats and does not provide a `<picture>` element, nor does it support `srcset`.

Use the HTML `<img>` element when you cannot use the `<Image />` component, for example:

>   - for unsupported image formats, like `.svg`

Reviewer 2: You can pass `.svg` to `<Image>`, they won't be processed but you still get the benefits regarding CLS, loading/decoding and alt.

>   - for unsupported image formats, like `.svg`

Author: ```suggestion
  - for unsupported image formats
```

Image CDNs work with all Astro image options. Use an image's full URL as the `src` attribute in the `<Image />` component, an `<img>` tag, or in Markdown notation. For image optimization with remote images, also [configure your authorized domains or URL patterns](#authorizing-remote-images).

Alternatively, if the CDN provides a Node.js SDK, you can use that in your project. For example, [Cloudinary’s SDK](https://cloudinary.com/documentation/node_integration) can generate an `<img>` tag with the appropriate `src` for you.

## Images in Markdown files

Use standard Markdown `![alt](src)` syntax in your `.md` files. This syntax works with Astro's Image Services API to optimize your local images and authorized remote images.

```md
<!-- src/pages/post-1.md -->

# My Markdown Page

<!-- Local image stored in src/assets/ -->
<!-- Use a relative file path or import alias -->
![A starry night sky.](../assets/stars.png)

<!-- Image stored in public/images/ -->
<!-- Use the file path relative to public/ -->
![A starry night sky.](/images/stars.png)

<!-- Remote image on another server -->
<!-- Use the full URL of the image -->
![Astro](https://example.com/images/remote-image.png)
```

> The `<img>` tag is not supported, nor is the `<Image />` component.

Reviewer 2: `<img>` does work in Markdown, but images won't be resolved or processed, much like in `.astro` files

> If you require more control over your image attributes, we recommend using the `.mdx` file format, which allows you to include Astro's `<Image />` component or a JSX `<img />` tag in addition to the Markdown syntax. Use the [MDX integration](/en/guides/integrations-guide/mdx/) to add support for MDX to Astro.

Reviewer 2: In MDX, `<img>` also isn't processed

---
title: My Page title
---
import { Image } from 'astro:assets';
import rocket from '../assets/rocket.png';

# My MDX Page

// Local image stored in src/assets/
<Image src={rocket} alt="A rocketship in space."/>
<img src={rocket.src} alt="A rocketship in space." />
![A rocketship in space](../assets/rocket.png)

// Image stored in public/images/
<Image src="/images/stars.png" alt="A starry night sky." />
<img src="/images/stars.png" alt="A starry night sky." />
![A starry night sky.](/images/stars.png)

// Remote image on another server
<Image src="https://example.com/images/remote-image.png" />
<img src="https://example.com/images/remote-image.png" />
![Astro](https://example.com/images/remote-image.png)

```

## Images in Content Collections

You can declare an associated image for a content collections entry, such as a blog post's cover image, in your frontmatter using its path relative to the current folder:

```md title="src/content/blog/my-post.md" {3}
---
title: "My first blog post"
cover: "./firstpostcover.jpeg" # will resolve to "src/content/blog/firstblogcover.jpeg"
coverAlt: "A photograph of a sunset behind a mountain range."
---

This is a blog post
```

> The `image` helper for content collections lets you validate the image metadata using Zod.

Reviewer 2: This schema is actually required for the image to be resolved, otherwise it's just treated as a normal string. It being able to validate the image dimensions is a bonus

> The `image` helper for content collections lets you validate the image metadata using Zod.

Author: ```suggestion
The `image` helper for the content collections schema lets you validate the image metadata using Zod.
```

Image CDNs work with all Astro image options. Use an image's full URL as the `src` attribute in the `<Image />` component, an `<img>` tag, or in Markdown notation. For image optimization with remote images, also [configure your authorized domains or URL patterns](#authorizing-remote-images).

Alternatively, if the CDN provides a Node.js SDK, you can use that in your project. For example, [Cloudinary’s SDK](https://cloudinary.com/documentation/node_integration) can generate an `<img>` tag with the appropriate `src` for you.

## Images in Markdown files

Use standard Markdown `![alt](src)` syntax in your `.md` files. This syntax works with Astro's Image Services API to optimize your local images and authorized remote images.

```md
<!-- src/pages/post-1.md -->

# My Markdown Page

<!-- Local image stored in src/assets/ -->
<!-- Use a relative file path or import alias -->
![A starry night sky.](../assets/stars.png)

<!-- Image stored in public/images/ -->
<!-- Use the file path relative to public/ -->
![A starry night sky.](/images/stars.png)

<!-- Remote image on another server -->
<!-- Use the full URL of the image -->
![Astro](https://example.com/images/remote-image.png)
```

> The `<img>` tag is not supported, nor is the `<Image />` component.

Author: ```suggestion
The `<img>` tag is not supported for local images,  and the `<Image />` component is unavailable in `.md` files.
```

As an alternative, you can wrap these components in another Astro component for reuse. For example, you could create a component for your blog post images:

```astro title="src/components/BlogPostImage.astro"
---
import { Image } from 'astro:assets';

const {src, ...attrs} = Astro.props;
---
<Image src={src} {...attrs} />

<style>
  img :global(img), svg {
    margin-block: 2.5rem;
    border-radius: 0.75rem;
  }
</style>
```

### `<img>`

The [Astro template syntax](/en/core-concepts/astro-syntax/) also supports writing an `<img>` tag directly, with full control over its final output. These images will not be processed and optimized.

It accepts all HTML `<img>` tag properties, and the only required property is `src`.

#### Local images in `src/`

Local images must be **imported from the relative path** from the existing `.astro` file, or configure and use an [import alias](/en/guides/aliases/). Then, you can access the image's `src` and other properties to use in the `<img>` tag.

For example, use the image's own `height` and `width` properties by default to avoid CLS and improve Core Web Vitals.

> ```astro "myDog.src"

Reviewer 3: Since this is an example of using `height` and `width`, I think these should be highlighted here.

```suggestion
```astro "myDog.width" "myDog.height"
```

    Remove any component attributes that are not [currently supported image asset properties](/en/guides/images/#properties).

    For example, `aspectRatio` is no longer supported, as it is now automatically inferred from the `width` and `height` attributes.

      ```astro title="src/components/MyComponent.astro" del= {2,11} ins={3}
      ---
      import { Image } from '@astrojs/image/components';
      import { Image } from 'astro:assets'
      import localImage from "../assets/logo.png";
      const localAlt = "The Astro Logo";
      ---

      <Image
        src={localImage}
        width={300}
        aspectRatio="16:9"
        alt={localAlt}
      />
      ```

3. Remove any existing `<Picture />` components.

    Currently, the built-in assets feature does not include a `<Picture />` component.

    Instead, you can use the HTML image attributes `srcset` and `sizes` or the `<picture>` tag [for art direction or to create responsive images](https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images#art_direction).

> 4. Uninstall Sharp.
> 
>     [Sharp](https://github.com/lovell/sharp) is now the default image service used for `astro:assets`. You no longer need to install the dependency locally in your project.
> 
>     Uninstall Sharp using your package manager, or by removing manually from your `package.json`.
> 

Author: ```suggestion
4. Choose a default image service.

    [Sharp](https://github.com/lovell/sharp) is now the default image service used for `astro:assets`. If you would like to use Sharp, no configuration is required.
```

However, both of these properties are required for remote images and images stored in your `public/` folder as Astro is unable to analyze these files.

##### format 

You can optionally state the [image file type](https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Image_types#common_image_file_types) output to be used.

By default, the `<Image />` component will produce a `.webp` file.

##### quality

`quality` is an optional property that can either be:
- a preset (`low`, `mid`, `high`, `max`) that is automatically normalized between formats.
- a number from `0` to `100` (interpreted differently between formats).

##### Additional properties

In addition to the properties above, the `<Image />` component accepts all properties accepted by the HTML `<img>` tag.

For example, you can provide a `class` to the final `<img>` element.

```astro title="src/pages/index.astro"
---
import { Image } from 'astro:assets';
import myImage from "../assets/my_image.png";
---

<!-- `alt` is mandatory on the Image component -->
<Image src={myImage} alt="" class="my-class" />
```

> ```astro
> <!-- Output -->
> <img
>   src="/_astro/my_image.hash.webp"
>   width="1600"
>   height="900"
>   decoding="async"
>   loading="lazy"
>   class="my-class"
>   alt=""
> />
> ```

Reviewer 4: ``````suggestion
```html
<!-- Output -->
<img
  src="/_astro/my_image.hash.webp"
  width="1600"
  height="900"
  decoding="async"
  loading="lazy"
  class="my-class"
  alt=""
/>
```
``````

			message: "Cover image must be at least 1080 pixels wide!",
		}),
		coverAlt: z.string(),
	}),
});

export const collections = {
	blog: blogCollection,
};
```

The image will be imported and transformed into metadata, allowing you to pass it as a `src` to `<Image/>`, `<img>`, or `getImage()`. 

The example below shows a blog index page that renders the cover photo and title of each blog post from the schema above:

```astro title="src/pages/blog.astro" {10}
---
import { Image } from "astro:assets";
import { getCollection } from "astro:content";
const allBlogPosts = await getCollection("blog");
---

{
	allBlogPosts.map((post) => (
		<div>
			<Image src={post.data.cover} alt={post.data.coverAlt} />
			<h2>
				<a href={"/blog/" + post.slug}>{post.data.title}</a>
			</h2>
		</div>
	))
}
```

## Images in UI framework components

When adding images in a UI framework component, use the framework's own image syntax to render an image (e.g. `<img />` in JSX, `<img>` in Svelte).

Local images must **first be imported** to access their [image properties](#local-images-in-src) such as `src`.

```jsx title="src/components/ReactImage.jsx"
import stars from "../assets/stars.png"

> export default function ReactImgage () {

Reviewer: ```suggestion
export default function ReactImage() {
```

export default defineConfig({
  image: {
    service: squooshImageService(),
  },
});
```

## Community Integrations

There are several third-party [community image integrations](https://astro.build/integrations?search=images) for optimizing and working with images in your Astro project.

## Upgrade to v3.0 from v2.x

`astro:assets` is no longer behind an experimental flag in Astro v3.0.

`<Image />` is now a built-in component and the previous `@astrojs/image` integration has been removed.

These and other accompanying changes to using images in Astro may cause some breaking changes when you upgrade your Astro project from an earlier version.

Please follow the instructions below as appropriate to upgrade an Astro v2.x project to v3.0. 

### Upgrade from `experimental.assets`

If you had previously enabled the experimental flag for `astro:assets`, you will need to update your project for Astro v3.0 which now includes assets features by default.

#### Remove `experimental.assets` flag

Remove the experimental flag:

```js title="astro.config.mjs" del={4-6}
import { defineConfig } from 'astro/config';

export default defineConfig({
  experimental: {

>    assets: true

Reviewer 3: Fix indentation:

```suggestion
    assets: true
```

```

#### Remove the `~/assets` import alias

This import alias is no longer included by default with `astro:assets`. If you were using this alias with experimental assets, you must convert them to relative file paths, or [create your own import aliases](/en/guides/aliases/).

```astro title="src/pages/posts/post-1.astro" del={2} ins={3}
---
import rocket from '~/assets/rocket.png'
import rocket from '../../assets/rocket.png';
---
```

#### Add simple asset support for Cloudflare, Deno, Vercel Edge and Netlify Edge

 Astro v3.0 allows `astro:assets` to work without errors in Cloudflare, Deno, Vercel Edge and Netlify Edge, which do not support Astro's built-in Squoosh and Sharp image optimization. Note that Astro does not perform any image transformation and processing in these environments. However, you can still enjoy the other benefits of using `astro:assets`, including no Cumulative Layout Shift (CLS), the enforced `alt` attribute, and a consistent authoring experience.

 If you previously avoided using `astro:assets` because of these constraints, you can now use them without issues. You can configure the no-op image service to explicitly opt-in to this behavior:

>  ```astro
>  // astro.config.mjs
>  export default {
>    image: {
>      service: {
>        entrypoint: 'astro/assets/services/noop'
>      }
>    }
>  }
>  ```

Reviewer 3: Use the `defineConfig` function here and treat this file as a js file:

````suggestion
```js title="astro.config.mjs" ins={4-8}
import { defineConfig } from 'astro/config';

export default defineConfig({
  image: {
    service: {
      entrypoint: 'astro/assets/services/noop'
    }
  }
});
```
````

### Decide where to store your images

> See the Images guide to help you decide [where to store your images](/en/guides/images/#where-to-store-images). You may wish to take advantage of new options for storing your images with the added flexibility `astro:assets` brings. For example, relative images from your project `src/` can now be referenced in Markdown, MDX, and Markdoc using standard Markdown `![alt](src)` syntax.

Reviewer 3: ```suggestion
See the Images guide to help you decide [where to store your images](#where-to-store-images). You may wish to take advantage of new options for storing your images with the added flexibility `astro:assets` brings. For example, relative images from your project `src/` can now be referenced in Markdown, MDX, and Markdoc using standard Markdown `![alt](src)` syntax.
```

### Update existing `<img>` tags

> Previously, importing an image would return a simple `string` with the path of the image. Now, imported image assets now match the following signature:

Reviewer 3: Remove the redundant "now":

```suggestion
Previously, importing an image would return a simple `string` with the path of the image. Now, imported image assets match the following signature:
```

```ts
interface ImageMetadata {
  src: string;
  width: number;
  height: number;
  format: string;
}
```

You must update the `src` attribute of any existing `<img>` tags (including any [images in UI framework components](#images-in-ui-framework-components)) and you may also update other attributes that are now available to you from the imported image.

```astro title="src/components/MyComponent.astro" ".src" ".width" ".height" del={4} ins={6}
---
import rocket from '../images/rocket.svg';
---
<img src={rocket} width="250" height="250" alt="A rocketship in space." />

<img src={rocket.src} width={rocket.width} height={rocket.height} alt="A rocketship in space." />
```

### Update your Markdown, MDX, and Markdoc files

Relative images from your project `src/` can now be referenced in Markdown, MDX, and Markdoc using standard Markdown `![alt](src)` syntax.

> This allows you the option to move your images from the `public/` directory to your project `src/` where they will now be processed and optimized. Your existing images in `public/` and remote images are still valid but are not optimized by Astro's build process.

Reviewer 3: ```suggestion
This allows you to move your images from the `public/` directory to your project `src/` where they will now be processed and optimized. Your existing images in `public/` and remote images are still valid but are not optimized by Astro's build process.
```

(Or something like "This gives you the option to move"?)

    ```js del={3,7}
    // astro.config.mjs
    import { defineConfig } from 'astro/config';
    import image from '@astrojs/image';

    export default defineConfig({
      integrations: [
        image(),
      ]
    })
    ```

2. Migrate any existing `<Image />` components.

    Change all `import` statements from `@astrojs/image/components` to `astro:assets` in order to use the new built-in `<Image />` component.

    Remove any component attributes that are not [currently supported image asset properties](/en/guides/images/#properties).

    For example, `aspectRatio` is no longer supported, as it is now automatically inferred from the `width` and `height` attributes.

      ```astro title="src/components/MyComponent.astro" del= {2,11} ins={3}
      ---
      import { Image } from '@astrojs/image/components';
      import { Image } from 'astro:assets'
      import localImage from "../assets/logo.png";
      const localAlt = "The Astro Logo";
      ---

      <Image
        src={localImage}
        width={300}
        aspectRatio="16:9"
        alt={localAlt}
      />
      ```

> 3. Remove any existing `<Picture />` components

Reviewer 3: Add a period at the end of the sentence for consistency:

```suggestion
3. Remove any existing `<Picture />` components.
```

    Currently, the built-in assets feature does not include a `<Picture />` component.

    Instead, you can use the HTML image attributes `srcset` and `sizes` or the `<picture>` tag [for art direction or to create responsive images](https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images#art_direction).

> 4. Uninstall Sharp

Reviewer 3: Same as above:

```suggestion
4. Uninstall Sharp.
```
